js/app.js

```diff
@@ -0,0 +1,128 @@
+const todoListElement = document.querySelector('#todo-list')
+const todoCountElement = document.querySelector('.count-container')
+
+const store = createStore(reducer);
+
+store.subscribe(() => {
+	const {todoList, filter} = store.getState();
+	const filteredList = Utils.todoListFilter(todoList, filter);
+	todoListElement.innerHTML = TodoList({list: filteredList})
+	todoCountElement.innerHTML = TodoFooter({list: filteredList, filter})
+});
+
+const localSave = () => {
+	const {id, todoList} = store.getState();
+	Utils.setStorageItem('id', id);
+	Utils.setStorageItem('todoList', todoList)
+}
+
+const addTodo = (text) => {
+	store.dispatch(actionCreator(ADD_TODO, text))
+	localSave();
+}
+const removeTodo = (id) => {
+	store.dispatch(actionCreator(REMOVE_TODO, id))
+	localSave();
+}
+const toggleTodoDone = (id) => {
+	store.dispatch(actionCreator(TOGGLE_TODO_DONE, id));
+	localSave();
+}
+
+const onTodoEdit = (id) => {
+	store.dispatch(actionCreator(TOGGLE_TODO_EDIT, id));
+	const input = todoListElement.querySelector('.editing .edit');
+	input.focus();
+	const val = input.value;
+	input.value = '';
+	input.value = val;
+	input.addEventListener('blur', () => {
+		toggleTodoEdit(id)
+	})
+}
+
+const toggleTodoEdit = (id) => {
+	store.dispatch(actionCreator(TOGGLE_TODO_EDIT, id));
+	const input = todoListElement.querySelector('.editing .edit');
+	input?.focus();
+	input?.addEventListener('blur', () => {
+		toggleTodoEdit(id)
+	})
+}
+const setTodo = (id, text) => {
+	store.dispatch(actionCreator(SET_TODO, {id, text}));
+	localSave();
+}
+const setFilter = (filter) => store.dispatch(actionCreator(SET_FILTER, filter));
+const initTodoApp = ({id, todoList, filter}) => store.dispatch(actionCreator(INIT_TODO_APP, {
+	id,
+	todoList,
+	filter
+}));
+
+window.addEventListener('load', () => {
+	initTodoApp({
+		id: Utils.getStorageItem('id') || 0,
+		todoList: Utils.getStorageItem('todoList') || [],
+		filter: window.location.hash?.slice(1) || 'all'
+	})
+})
+
+todoInputElement.addEventListener('keyup', (event) => {
+	const {key, target: {value}} = event;
+	const trimValue = value.trim();
+	if (key === "Enter" && trimValue) {
+		addTodo(trimValue)
+		event.target.value = '';
+	}
+})
+
+window.addEventListener('hashchange', () => {
+	const filter = window.location.hash?.slice(1) || "all";
+	setFilter(filter)
+})
+
+todoListElement.addEventListener('click', ({target: {dataset: {id, role}}}) => {
+	switch (role) {
+		case 'toggle':
+			toggleTodoDone(+id)
+			return;
+		case 'remove':
+			removeTodo(+id)
+			return;
+	}
+})
+
+todoListElement.addEventListener('dblclick', ({target: {dataset: {id, role}}}) => {
+	if (role === 'edit') {
+		onTodoEdit(+id)
+	}
+})
+
+todoListElement.addEventListener('keyup', ({key, target: {value, dataset: {id, role}}}) => {
+	if (role === 'change') {
+		switch (key) {
+			case 'Escape':
+				toggleTodoEdit(+id)
+				return;
+			case 'Enter':
+				setTodo(+id, value)
+				return;
+		}
+	}
```

Comment:
```suggestion
	if (!role === 'change') return;
	switch (key) {
		case 'Escape':
			return toggleTodoEdit(+id);
		case 'Enter':
			return setTodo(+id, value);
	}
```
- 가능하면 if로 인하여 코드의 depth가 깊어지는건 피하는게 좋습니다!
- 실행=>return이 아니라 실행한걸 바로 return 해버리면됩니다. 리턴 값이 없으면 어차피 undefined 형태로 반환하니까요!